File: module_common/src/main/java/module/common/utils/StatusBarUtils.java
package module.common.utils;

import android.app.Activity;
import android.graphics.Rect;
import android.view.View;
import android.view.ViewGroup;

/**
 * @describe: 状态栏工具类
 * @date: 2019/12/31
 * @author: Mr Bai
 */
public class StatusBarUtils {


    /**
     * 获取状态栏高度
     *
     * @return > 0 success; <= 0 fail
     */
    public static int getStatusHeight(Activity activity) {
        int statusHeight = 0;
        Rect localRect = new Rect();
        activity.getWindow().getDecorView().getWindowVisibleDisplayFrame(localRect);
        statusHeight = localRect.top;
        if (0 == statusHeight) {
            Class<?> localClass;
            try {
                localClass = Class.forName("com.android.internal.R$dimen");
                Object localObject = localClass.newInstance();
                int i5 = Integer.parseInt(localClass
                        .getField("status_bar_height").get(localObject)
                        .toString());
                statusHeight = activity.getResources().getDimensionPixelSize(i5);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
        return statusHeight;
    }


    /**
     * @describe: 设置距离状态栏的高度
     * @date: 2019/12/31
     */
    public static void setMarginStatusBarHeight(Activity activity,View view){
        if(activity==null || view==null){
            return;
        }
        ViewGroup.MarginLayoutParams layoutParams = (ViewGroup.MarginLayoutParams) view.getLayoutParams();
        layoutParams.topMargin = getStatusHeight(activity);
        view.setLayoutParams(layoutParams);
    }
}

File: module_common/src/main/java/module/common/utils/StatusBarUtils.kt
package module.common.utils

import android.app.Activity
import android.os.Build
import android.view.View
import android.view.ViewGroup.MarginLayoutParams
import android.view.WindowInsets
import kotlin.math.abs

/**
 * @describe: 状态栏工具类
 * @date: 2019/12/31
 * @author: Mr Bai
 */
object StatusBarUtils {
    /**
     * 获取状态栏高度
     *
     * @return > 0 success; <= 0 fail
     */
    fun getStatusHeight(activity: Activity): Int {
        var height = 0
        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.R){
            val decorView: View = activity.window.decorView
            decorView.rootWindowInsets?.let {
                val insets = it.getInsets(WindowInsets.Type.statusBars())
                height = abs(insets.bottom - insets.top)
            }
        }else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            val decorView: View = activity.window.decorView
            decorView.rootWindowInsets?.let {
                height = it.systemWindowInsetTop
            }
        } else {
            val resourceId = activity.resources.getIdentifier("status_bar_height", "dimen", "android")
            if (resourceId > 0) {
                height = activity.resources.getDimensionPixelSize(resourceId)
            }
        }
        LogUtils.printI("StatusHeight=$height")
        return height
    }

    /**
     * @describe: 设置距离状态栏的高度
     * @date: 2019/12/31
     */
    fun setMarginStatusBarHeight(activity: Activity?, view: View?) {
        if (activity == null || view == null) {
            return
        }
        var statusHeight = getStatusHeight(activity)
        if(statusHeight == 0){
            statusHeight = SPHelper.get(activity.applicationContext,SPHelper.STATUS_HEIGHT,0) as Int
        }

        val layoutParams = view.layoutParams as MarginLayoutParams
        layoutParams.topMargin = statusHeight
        view.layoutParams = layoutParams
    }

}